python/mxnet/gluon/data/dataset.py

```diff
@@ -84,14 +84,13 @@ def shard(self, num_shards, index):
         """
         assert index < num_shards, 'Shard index of out bound: %d out of %d'%(index, num_shards)
         assert num_shards > 0, 'Number of shards must be greater than 0'
+        assert index >= 0, 'Index must be non-negative'
         length = len(self)
-        shard_len = length // num_shards
+        shard_len = (length + num_shards - 1) // num_shards
         # Compute the start index for this partition
         start = shard_len * index
         # Compute the end index for this partition
-        end = start + shard_len
-        if index == num_shards - 1:
-            end = length
+        end = start + shard_len if index < num_shards - 1 else length
```

Comment:
When `length=6, num_shareds=4`, 
`shared_len = (6 + 4 - 1) // 4 = 2`

The interval of the four partitions are `[0, 2), [2, 4), [4, 6), [6, 6)`

Reply:
Got it. Thanks a lot to point this out! I'll fix it